File: apps/briefcase-mobile-signer/android/app/src/main/java/com/credentialbriefcase/mobilesigner/BriefcasedClient.kt
package com.credentialbriefcase.mobilesigner

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.HttpUrl.Companion.toHttpUrl

class BriefcasedClient(baseUrl: String) {
    private val http = OkHttpClient()
    private val json = Json { ignoreUnknownKeys = true }
    private val base = baseUrl.toHttpUrl()

    suspend fun completePairing(pairingId: String, req: SignerPairCompleteRequest): SignerPairCompleteResponse {
        return post("/v1/signer/pair/$pairingId/complete", req)
    }

    suspend fun listApprovals(req: SignerSignedRequest): ListApprovalsResponse {
        return post("/v1/signer/approvals", req)
    }

    suspend fun approve(approvalId: String, req: SignerSignedRequest): ApproveResponse {
        return post("/v1/signer/approvals/$approvalId/approve", req)
    }

    private suspend inline fun <reified Req, reified Res> post(path: String, body: Req): Res {
        val media = "application/json; charset=utf-8".toMediaType()
        val bodyJson = json.encodeToString(body)
        val url = base.resolve(path) ?: throw IllegalArgumentException("bad url")

        val request = Request.Builder()
            .url(url)
            .post(bodyJson.toRequestBody(media))
            .build()

        return withContext(Dispatchers.IO) {
            http.newCall(request).execute().use { resp ->
                val bytes = resp.body?.bytes() ?: ByteArray(0)
                val text = bytes.toString(Charsets.UTF_8)
                if (!resp.isSuccessful) {
                    val err = runCatching { json.decodeFromString<ErrorResponse>(text) }.getOrNull()
                    throw IllegalStateException(err?.let { "${it.code}: ${it.message}" } ?: "HTTP ${resp.code}")
                }
                json.decodeFromString<Res>(text)
            }
        }
    }
}

File: apps/briefcase-mobile-signer/ios/BriefcaseMobileSigner/Core/BriefcasedClient.swift
import Foundation

final class BriefcasedClient {
    private let baseURL: URL
    private let session: URLSession

    init(baseURL: URL, session: URLSession = .shared) {
        self.baseURL = baseURL
        self.session = session
    }

    func completePairing(pairingId: UUID, req: SignerPairCompleteRequest) async throws -> SignerPairCompleteResponse {
        try await post("/v1/signer/pair/\(pairingId.uuidString)/complete", body: req)
    }

    func listApprovals(req: SignerSignedRequest) async throws -> ListApprovalsResponse {
        try await post("/v1/signer/approvals", body: req)
    }

    func approve(approvalId: UUID, req: SignerSignedRequest) async throws -> ApproveResponse {
        try await post("/v1/signer/approvals/\(approvalId.uuidString)/approve", body: req)
    }

    private func post<Req: Encodable, Res: Decodable>(_ path: String, body: Req) async throws -> Res {
        guard let url = URL(string: path, relativeTo: baseURL)?.absoluteURL else {
            throw URLError(.badURL)
        }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await session.data(for: req)
        guard let http = response as? HTTPURLResponse else {
            throw URLError(.badServerResponse)
        }
        if (200 ..< 300).contains(http.statusCode) {
            return try JSONDecoder().decode(Res.self, from: data)
        }
        if let err = try? JSONDecoder().decode(ErrorResponse.self, from: data) {
            throw err
        }
        throw URLError(.badServerResponse)
    }
}
